keyboards/zsa/voyager/voyager.c

```diff
@@ -0,0 +1,359 @@
+                keyboard_config.led_level ^= 1;
+                eeconfig_update_kb(keyboard_config.raw);
+                if (keyboard_config.led_level) {
+                    layer_state_set_kb(layer_state);
+                } else {
+                    STATUS_LED_1(false);
+                    STATUS_LED_2(false);
+                    STATUS_LED_3(false);
+                    STATUS_LED_4(false);
+                }
+            }
+            break;
+#endif
+#ifdef RGB_MATRIX_ENABLE
+        case TOGGLE_LAYER_COLOR:
+            if (record->event.pressed) {
+                keyboard_config.disable_layer_led ^= 1;
+                if (keyboard_config.disable_layer_led) rgb_matrix_set_color_all(0, 0, 0);
+            }
+            break;
+        case RGB_TOG:
+            if (record->event.pressed) {
+                switch (rgb_matrix_get_flags()) {
+                    case LED_FLAG_ALL: {
+                        rgb_matrix_set_flags(LED_FLAG_NONE);
+                        rgb_matrix_set_color_all(0, 0, 0);
+                    } break;
+                    default: {
+                        rgb_matrix_set_flags(LED_FLAG_ALL);
+                    } break;
+                }
+            }
+            return false;
+#endif
+    }
+    return true;
+}
+
+void matrix_init_kb(void) {
+    keyboard_config.raw = eeconfig_read_kb();
+
+    if (!keyboard_config.led_level && !keyboard_config.led_level_res) {
+        keyboard_config.led_level     = true;
+        keyboard_config.led_level_res = 0b11;
+        eeconfig_update_kb(keyboard_config.raw);
+    }
+    matrix_init_user();
+}
+
+void eeconfig_init_kb(void) { // EEPROM is getting reset!
+    keyboard_config.raw           = 0;
+    keyboard_config.led_level     = true;
+    keyboard_config.led_level_res = 0b11;
+    eeconfig_update_kb(keyboard_config.raw);
+    eeconfig_init_user();
+}
+
+
+// The keyboard uses a dfu-util compatible custom bootloader.
+// Rather than implementing in core, just override relevant functions.
+
+#define APP_ADDRESS 0x08002000
+
+void bootloader_jump(void) {
+    // The ignition bootloader is checking for a high signal on A8 for 100ms when powering on the board.
+    // Setting both A8 and A9 high will charge the capacitor quickly.
+    // Setting A9 low before reset will cause the capacitor to discharge
+    // thus making the bootloder unlikely to trigger twice between power cycles.
+    setPinOutputPushPull(A9);
+    setPinOutputPushPull(A8);
+    writePinHigh(A9);
+    writePinHigh(A8);
+    wait_ms(500);
+    writePinLow(A9);
+
+    NVIC_SystemReset();
+}
+
+
+void mcu_reset(void) {
+    // When resetting the MCU, we want to jump to the application.
+    SCB->AIRCR = APP_ADDRESS & 0xFFFF;
+
+    // Set the stack pointer to the applications stack pointer
+    __asm__ volatile("msr msp, %0" ::"g"(*(volatile uint32_t *)APP_ADDRESS));
+
+    // Jump to the application
+    (*(void (**)())(APP_ADDRESS + 4))();
+    while (1)
+        ;
```

Comment:
Not sure whether jumping to the app like that without going through an actual system reset would work reliably in all cases (although ChibiOS apparently resets everything it can through RCC, maybe there are some differences from the real system reset).  Maybe you would be able to force A8 low and then do `NVIC_SystemReset()`, going through the custom bootloader as for the initial startup?

If you keep the current code, you may also need to do things like `enter_bootloader_mode_if_requested()` does to disable all interrupts and clear all pending interrupts to avoid getting any problems if some interrupt happens during the restart.

Reply:
It looks like a lot of this is gd32 specific.  The A8 low and nvic sysreset seems to work reliably.